common4j/src/main/com/microsoft/identity/common/java/cache/SharedPreferencesAccountCredentialCacheWithMemoryCache.java

```diff
@@ -332,6 +339,10 @@ public boolean test(String value) {
                 // Remove the entry and return null...
                 Logger.warn(methodTag, "The returned Credential was uninitialized. Removing...");
                 mSharedPreferencesFileManager.remove(cacheKey);
+            } else if (AccessTokenRecord.class == clazz
+                    && !mSha1Cleared
+                    && isSha1ApplicationIdentifierAccessToken(credential)) {
+                    mSharedPreferencesFileManager.remove(cacheKey);
             }
```

Comment:
For the sake of simplicity, I don't think it would be a perf problem to implement this load->cleanup->save logic onc as something like `cleanUpSha1CredentialsIfNeeded` in AbstractAccountCredentialCache using the available getCredentials() abstraction.  SPACC could call that function in its constructor, basically where you have it in the block now (assuming we decide its okay to do it then for perf).  And SPACCWMC could call it as the last line of its load() method that's running synchronously on a background thread.

Reply:
This is a great catch.  I forgot that implementing this in terms of getCredentials() at the high level would have these locking problems since they're inherently public methods that need to guard to wait for the load.  This solution works for me.  One more small suggestion now that we're doing things this way: can you get rid of the finally block and instead just have the duplicate mLoaded=true/mCacheLock.notifyall() inside the catch block? That should cover both the success and exception case.  Now that we need to handle success specially, the finally block with an if makes less sense than just putting it in the catch.

Reply:
welp... ended up keeping the finally block due to going with your original suggestion, as mentioned in my other comment. 😅
commit with new changes: https://github.com/AzureAD/microsoft-authentication-library-common-for-android/commit/4b24755cb9168aa44753751b7b993bf65e2f585f